sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java

```diff
@@ -51,56 +53,37 @@
 import org.apache.beam.sdk.values.TupleTag;
 import org.apache.beam.sdk.values.ValueInSingleWindow;
 import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
- * A harness for unit-testing a {@link DoFn}.
- *
- * <p>For example:
- *
- * <pre> {@code
- * DoFn<InputT, OutputT> fn = ...;
- *
- * DoFnTester<InputT, OutputT> fnTester = DoFnTester.of(fn);
- *
- * // Set arguments shared across all bundles:
- * fnTester.setSideInputs(...);      // If fn takes side inputs.
- * fnTester.setOutputTags(...);  // If fn writes to more than one output.
- *
- * // Process a bundle containing a single input element:
- * Input testInput = ...;
- * List<OutputT> testOutputs = fnTester.processBundle(testInput);
- * Assert.assertThat(testOutputs, Matchers.hasItems(...));
- *
- * // Process a bigger bundle:
- * Assert.assertThat(fnTester.processBundle(i1, i2, ...), Matchers.hasItems(...));
- * } </pre>
- *
- * @param <InputT> the type of the {@link DoFn}'s (main) input elements
- * @param <OutputT> the type of the {@link DoFn}'s (main) output elements
+ * @deprecated Instead of this, write a {@link TestPipeline} that verifies expectations using {@link
+ *     PAssert} and run it with the {@code DirectRunner} (or other runners, as applicable). For
+ *     simple cases, use {@link Create} to provide input. For complex cases involving triggers, use
+ *     {@link TestStream} for input data and behavior.
  */
+@Deprecated
 public class DoFnTester<InputT, OutputT> implements AutoCloseable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DoFnTester.class);
+
   /**
-   * Returns a {@code DoFnTester} supporting unit-testing of the given
-   * {@link DoFn}. By default, uses {@link CloningBehavior#CLONE_ONCE}.
-   *
-   * <p>The only supported extra parameter of the {@link DoFn.ProcessElement} method is
-   * {@link BoundedWindow}.
+   * @deprecated Use {@link TestStream} with the {@code DirectRunner}.
    */
   @SuppressWarnings("unchecked")
+  @Deprecated
   public static <InputT, OutputT> DoFnTester<InputT, OutputT> of(DoFn<InputT, OutputT> fn) {
     checkNotNull(fn, "fn can't be null");
+    LOG.warn(
+        "Your tests use DoFnTester, which may not exercise DoFns correctly. "
+            + "Please use TestStream instead.");
```

Comment:
I'd suggest to rephrase all "use TestStream" in this PR to "use TestPipeline".

Reply:
Agreed and done.